repository/src/main/java/org/alfresco/repo/domain/schema/script/DeleteNotExistsV3Executor.java

```diff
@@ -0,0 +1,437 @@
+            PreparedStatement deletePrepStmt, Set<Long> deleteIds, String primaryTableName, String primaryColumnName,
+            Pair<String, String>[] tableColumn) throws SQLException
+    {
+        int rowsProcessed = 0;
+        Long primaryId = null;
+        Long minSecId = 0L;
+        Long maxSecId = 0L;
+
+        // Get the potential IDS to delete from the primary table
+        Set<Long> potentialIdsToDelete = new HashSet<Long>();
+        try (ResultSet resultSet = primaryPrepStmt.getResultSet())
+        {
+            while (resultSet.next())
+            {
+                primaryId = resultSet.getLong(primaryColumnName);
+                potentialIdsToDelete.add(primaryId);
+            }
+        }
+
+        if (potentialIdsToDelete.size() == 0)
+        {
+            // Nothing more to do
+            return primaryId;
+        }
+
+        Long minPotentialId = Collections.min(potentialIdsToDelete);
+        Long maxPotentialId = Collections.max(potentialIdsToDelete);
+        rowsProcessed = potentialIdsToDelete.size();
+        processedCounter = processedCounter + rowsProcessed;
+
+        // Get a combined list of the ids present in the secondary tables
+        Set<Long> secondaryResults = getSecondaryResults(secondaryPrepStmts, tableColumn, minPotentialId, maxPotentialId);
+
+        if (secondaryResults.size() > 0)
+        {
+            minSecId = Collections.min(secondaryResults);
+            maxSecId = Collections.max(secondaryResults);
+
+            // From our potentialIdsToDelete list, remote all non-eligible ids: any id that is in a secondary table or
+            // any ID past the last ID we were able to access in the secondary tables (maxSecId)
+            Iterator<Long> it = potentialIdsToDelete.iterator();
+            while (it.hasNext())
+            {
+                Long id = it.next();
+                if (id > maxSecId || secondaryResults.contains(id))
+                {
+                    it.remove();
+                }
+            }
+
+            // The next starting primary ID for the next batch will either be the next last id evaluated from the
+            // primary table or, in case the secondary queries did not get that far, the last secondary table id
+            // evaluated (maxSecId)
+            primaryId = primaryId < maxSecId ? primaryId : maxSecId;
+        }
+
+        if (potentialIdsToDelete.size() > 0)
+        {
+            deleteInBatches(potentialIdsToDelete, deleteIds, primaryTableName, deletePrepStmt);
+        }
+
+        if (logger.isDebugEnabled())
+        {
+            logger.debug("RowsProcessed " + rowsProcessed + " from primary table " + primaryTableName + ". Primary: ["
+                    + minPotentialId + "," + maxPotentialId + "] Secondary: [" + minSecId + "," + maxSecId + "] Total Deleted: "
+                    + deletedCount);
+        }
+
+        // Do we need to pause?
+        if (processedCounter >= pauseAndRecoverBatchSize)
```

Comment:
I couldn't see how the batching worked. I'm sure it does, because I know you've tested it, but this is really a comment about the clarity of the code/documentation.  From what I saw then it looked like all the rows have been processed at this point and so it doesn't matter if we pause or not.

Reply:
The counter is only reset after each pause. Every row in each batch is processed, but when we reach the end of each batch we verify if we need to mark the job to pause before starting the next batch due to the pauseAndRecoverBatchSize being reached. If the flag is set to true, we have a break (L176), we close everything, pause and then resume where we left off (L205). 

When doing massive deletes on the DB after a period of time we start seeing severe performance degradation on the DB that leads to aborted connections. The pause was implemented to prevent that. After X rows processed, we stop for Y amount of time to allow the DB to recover before continuing.

Reply:
I don't see what's controlling the batches though. This is what I think the flow looks like:

```
process
{
    processBatch
    {
        pauseAndRecover = false;
        if (hasResults)
        {
            while (hasResults && !isTimeoutExceeded())
            {
                processPrimaryTableResultSet
                {
                    if (processedCounter >= pauseAndRecoverBatchSize)
                    {
                        pauseAndRecover = true;
                    }
                }
            }
        }
        if (pauseAndRecover)
        {
            pauseAndRecoverJob
            processBatch
        }
    }
}
```

So `process` always calls `processBatch` once and the only pause can happen after the whole lot is processed, or `isTimeoutExceeded()` returns true.  Maybe the code calling `process` handles the batching?  Or maybe I just missed how the batch is created.

Reply:
Its in the processBatch. Completing your flow:

`
process
{
    processBatch
    {
        pauseAndRecover = false;
        if (hasResults)
        {
            while (hasResults && !isTimeoutExceeded())
            {
                primaryId = processPrimaryTableResultSet
                {
                    if (processedCounter >= pauseAndRecoverBatchSize)
                    {
                        pauseAndRecover = true;
                    }
                }
                
               // If the next primaryId is null or flag to pause is true, exit
               if (primaryId == null || pauseAndRecover)
               {
                        break;
               }
                
               // Prepare for next batch doing the queries starting on primaryId + 1 and execute the query for the next batch
               primaryPrepStmt.setLong(1, primaryId + 1);
               primaryPrepStmt.setLong(2, tableUpperLimits[0]);
               hasResults = primaryPrepStmt.execute();
            }
        }
        if (pauseAndRecover)
        {
            pauseAndRecoverJob
            processBatch from last primaryId
        }
    }
}`